connect_four.py

```diff
@@ -42,25 +42,48 @@ def __vertical_win(self, row: int, col: int, color: str) -> bool:
                         return True
                 else:
                     return False
-        
-        return False  #this return shouldn't be necessary
 
     def __horizontal_win(self, row: int, col: int, color: str) -> bool:
-        #TODO change logic to be # of same color on left + # of same color on right
-        adjacent_count = 0
-        for col in range(self.num_cols): 
-            if self.current[row][col] == color:
-                adjacent_count += 1
-                if adjacent_count >= self.win_req:
-                    return True
-            else:
-                    adjacent_count = 0
-        
-        return False
+
+        left_count, right_count = 0, 0
+
+        while col - 1 - left_count >= 0 and self.current[row][col - 1 - left_count] == color: #while there is a column to the left and that piece is of same color
+            left_count += 1
+
+        while col + 1 + right_count < self.num_cols and self.current[row][col + 1 + right_count] == color: #while there is a column to the right and that piece is of same color
+            right_count += 1 
+
+        return 1 + left_count + right_count >= self.win_req   #current new piece + # of same color on left + # of same color on right
 
     def __diagonal_win(self, row: int, col: int, color: str) -> bool:
-        return False
-        #TODO.........................................
+
+        #check upper left diagonal + lower right diagonal (negative slope)
+        up_left, down_right = 0, 0
```

Comment:
the only small thing I would suggest is to break up this method into two smaller methods, one to handle up-down diagonal and the other to handle down-up. You can either call both within the diagonal win or you can change the check win to something like:

__vertical_win(row, col, color) or __horizontal_win(row, col, color) or __diagonal_win_up(row, col, color)  or __diagonal_win_down(row, col, color)